### ISSUE TITLE
Add constraints to generated interfaces

### ISSUE THREAD

**Issue by frankpf** [SweetIQ/schemats#77]
Is it possible to add information on constraints to the generated interfaces?

As an example, this table:

```sql
CREATE TABLE users (
  id                 text         NOT NULL PRIMARY KEY,
  email           text         NOT NULL UNIQUE,
  pwd_hash   char(60) NOT NULL,
  deleted        boolean NOT NULL DEFAULT false,
  name           text
)
```

could generate something like this:

```
export interface users {
   id: { type: string, primary: true }
   email: { type: string, unique: true }
   pwd_hash: { type: string }
   deleted: { type: boolean,  default: false }
   name: { type: string | null }
}
```

I understand that this feature would be overkill to most users, so it could be hidden behind a flag  `--mode verbose` or something like that.

If there's interest in this feature, I'd be willing to work on it.

**Comment by abenhamdine** [comment]
Good idea IMHO, could be also a way to implement https://github.com/SweetIQ/schemats/issues/12#issuecomment-270928311

**Comment by xiamx** [comment]
Sounds good!

My only concern is that the new interface generation will not be compatible with the current way we typecheck query returns (see example https://github.com/SweetIQ/schemats#writing-code-with-typed-schema)

Maybe these information can take a different key, so instead we have two alternatives:

1. 
```typescript
export interface users {
   id: string
   _id_meta: {primary: true}
   email: string
   _email_meta: {unique: true, comment: 'somecomment'}
}
```

2.
```typescript
export interface users {
   id: string
   email: string
}

export interface usersMeta {
   id_meta: {primary: true}
   email_meta: {unique: true, comment: 'somecomment'}
}
```

**Comment by frankpf** [comment]
One problem with that is that it becomes more inconvenient for libraries to access the relevant keys.

As an example, I want to add a method `findOne` to SafeQL that accepts only primary/unique keys as arguments. Using the scheme above, the user would have to pass  `"_id_meta"` as an argument instead of `"id"`.

Here's what I suggest:

```typescript
type HasTypeKey<T> = {
    [K in keyof T]: {
        type: any
    }
}
type SimpleSchema<T extends HasTypeKey<T>> = {
    [K in keyof T] : T[K]['type']
}

export interface usersMeta {
   id: { type: string, primary: true }
   email: { type: string, unique: true }
}

export type users = SimpleSchema<usersMeta>
```

You can check it on [TS playground](https://www.typescriptlang.org/play/#src=type%20HasTypeKey%3CT%3E%20%3D%20%7B%0D%0A%20%20%20%20%5BK%20in%20keyof%20T%5D%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20type%3A%20any%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0Atype%20SimpleSchema%3CT%20extends%20HasTypeKey%3CT%3E%3E%20%3D%20%7B%0D%0A%20%20%20%20%5BK%20in%20keyof%20T%5D%20%3A%20T%5BK%5D%5B'type'%5D%0D%0A%7D%0D%0A%0D%0Aexport%20interface%20usersMeta%20%7B%0D%0A%20%20%20id%3A%20%7B%20type%3A%20string%2C%20primary%3A%20true%20%7D%0D%0A%20%20%20email%3A%20%7B%20type%3A%20string%2C%20unique%3A%20true%20%7D%0D%0A%7D%0D%0A%0D%0Aexport%20type%20users%20%3D%20SimpleSchema%3CusersMeta%3E%0D%0A%0D%0A%0D%0Aconst%20users%3A%20Array%3Cusers%3E%20%3D%20%5B%5D%0D%0A%0D%0Afor%20(const%20user%20of%20users)%20%7B%0D%0A%20%20%20%20let%20x%20%3D%20user.id%0D%0A%20%20%20%20let%20y%20%3D%20user.email%0D%0A%7D) and confirm that it still type-checks correctly.

Essentially, only the "verbose" interfaces are generated (containing all the necessary information), and we use mapped types to transform them to the basic interfaces.

This would keep the library compatible with the current API and give library authors full access to the type information.

**Comment by frankpf** [comment]
@abenhamdine Yes, IMO we should output as much information as possible (for library authors to use) while keeping the types simple for end users.

What do you think of something like this:

```typescript
export interface usersMeta {
   id: { type: string, primary: true, originalType: 'text' },
   pwd_hash: { type: string, originalType: 'char(60)' }
}
```

Although I don't know if simple strings are the best representations for the SQL types. An alternative would be:
```typescript
originalType: { type: 'text' }
originalType: { type: 'char', length: 60 }
```

The same goes for check constraints. I see two options: 

1. ```typescript
check: 'price > 0`
```
2. ```typescript
check: { price: { greaterThan: 0 } }
```

Option 2 seems better, although it would be harder to implement.

**Comment by xiamx** [comment]
> Here's what I suggest:
> 
> ```typescript
> type HasTypeKey<T> = {
>     [K in keyof T]: {
>         type: any
>     }
> }
> type SimpleSchema<T extends HasTypeKey<T>> = {
>     [K in keyof T] : T[K]['type']
> }
> 
> export interface usersMeta {
>    id: { type: string, primary: true }
>    email: { type: string, unique: true }
> }
> 
> export type users = SimpleSchema<usersMeta>
> ```

Nice! This is a very clean solution! With this solution, we can generate the table column type interfaces along with the metadata like primary and unique. The meta interfaces like `usersMeta` can be unexported by default; and is only exported with `--mode verbose` is passed in. This will ensure that the default behaviour stays consistent for the end user. 

> ```typescript
> check: { price: { greaterThan: 0 } }
> ```

Looks like option 2 would require some facility to parse SQL DDL. This will surely take more time to implement.  IMO, it would be fine to leave the parsing of SQL types to downstream libraries, since we are returning valid SQL types as string.

**Comment by xiamx** [comment]
Yeah looks like not everything is easily accessible. Not having MySQL support for now would be fine, someone interested in it will eventually make a PR :D 

Feel free to open a pull request when you are ready.
